lib/robots/viomi/capabilities/ViomiCarpetModeControlCapability.js

```diff
@@ -0,0 +1,47 @@
+const CarpetModeControlCapability = require("../../../core/capabilities/CarpetModeControlCapability");
+const stateAttrs = require("../../../entities/state/attributes");
+
+class ViomiCarpetModeControlCapability extends CarpetModeControlCapability {
+    /**
+     * This function returns the last set carpet turbo mode setting..
+     * Viomi does not have a command to retrieve the current setting, so we fail if it was not set recently.
+     *
+     * @abstract
+     * @returns {Promise<boolean>}
+     */
+    async isEnabled() {
+        const state = this.robot.state.getFirstMatchingAttributeByConstructor(stateAttrs.CarpetTurboStateAttribute);
+        if (state === null) {
+            throw new Error("Carpet turbo state cannot be remembered by Valetudo. " +
+                "If you don't remember what you set it to, set it again.");
+        }
+        return state.value === stateAttrs.CarpetTurboStateAttribute.VALUE.ENABLED;
+    }
+
+    /**
+     * @abstract
+     * @returns {Promise<void>}
+     */
+    async enable() {
+        this.robot.state.upsertFirstMatchingAttribute(new stateAttrs.CarpetTurboStateAttribute({
+            value: stateAttrs.CarpetTurboStateAttribute.VALUE.ENABLED
+        }));
+
+        // 0: off, 1: medium, 2: turbo
+        await this.robot.sendCommand("set_carpetturbo", [2], {});
+    }
+
+    /**
+     * @abstract
+     * @returns {Promise<void>}
+     */
+    async disable() {
+        this.robot.state.upsertFirstMatchingAttribute(new stateAttrs.CarpetTurboStateAttribute({
```

Comment:
As explained above, we can simply replace this with a dynamic property on the viomi vacuum implementation.

Disclaimer:
People coming from statically typed languages will probably cry in terror just from reading this suggestion, however personally, I'd argue that this is the point of using a dynamically-typed language such as JS because it lets you do stuff like this without being forced to refactor everything. IMO it's okay to do that for stuff that is self-contained with no side-effects, as long as other options are evaluated and of course code gets refactored if there are more than just a few hacks like these

Anyways:

This should be something like `this.robot.carpetModeEnabled = true` or maybe if there's more of these `this.robot.viomiHacks.carpetModeEnabled` or something like that

Reply:
I get your point about it not being state and I agree, as I said I don't think it's a good solution either. However... I also think that setting it as a dynamic property is even nastier. Anyway, your project, your style decisions :)